prov/tcp/src/xnet_rdm.c

```diff
@@ -695,13 +695,183 @@ static struct fi_ops_ep xnet_rdm_ep_ops = {
+	struct fi_mr_attr *attr = (struct fi_mr_attr *)node->data;
+	struct xnet_domain *subdomain = context;
+
+	ret = ofi_mr_map_insert(&subdomain->util_domain.mr_map, attr,
+				&attr->requested_key, attr->context,
+				((struct ofi_mr*)attr->context)->flags);
+	if (ret)
+		XNET_WARN_ERR(FI_LOG_MR, "ofi_mr_map_insert", ret);
+	else
+		ofi_atomic_inc32(&subdomain->util_domain.ref);
+}
+
+struct xnet_domain *find_subdomain(struct xnet_rdm *rdm)
+{
+	int i;
+	struct util_cntr *cntr;
+
+	if (!xnet_domain_multiplexed(&rdm->util_ep.rx_cq->domain->domain_fid)) {
+		return container_of(&rdm->util_ep.rx_cq->domain->domain_fid,
+				    struct xnet_domain, util_domain.domain_fid);
+	}
+
+	if (!xnet_domain_multiplexed(&rdm->util_ep.tx_cq->domain->domain_fid)) {
+		return container_of(&rdm->util_ep.tx_cq->domain->domain_fid,
+					 struct xnet_domain, util_domain.domain_fid);
+	}
+
+	for (i = 0; i < CNTR_CNT; i++) {
+		cntr = rdm->util_ep.cntrs[i];
+		if (!cntr)
+			continue;
+
+		if (!xnet_domain_multiplexed(&cntr->domain->domain_fid)) {
+			return container_of(&cntr->domain->domain_fid,
+					 struct xnet_domain, util_domain.domain_fid);
+		}
+	}
+
+	return NULL;
+}
+
+static void set_subdomain(struct xnet_rdm *rdm, struct xnet_domain *domain,
+			  struct xnet_domain *subdomain)
+{
+	int i;
+	struct util_cntr *cntr;
+
+	if (rdm->util_ep.rx_cq->domain == &domain->util_domain) {
+		ofi_atomic_dec32(&rdm->util_ep.rx_cq->domain->ref);
+		ofi_atomic_inc32(&subdomain->util_domain.ref);
+		rdm->util_ep.rx_cq->domain = &subdomain->util_domain;
+	}
+
+	if (rdm->util_ep.tx_cq->domain == &domain->util_domain) {
+		ofi_atomic_dec32(&rdm->util_ep.tx_cq->domain->ref);
+		ofi_atomic_inc32(&subdomain->util_domain.ref);
+		rdm->util_ep.tx_cq->domain = &subdomain->util_domain;
+	}
+
+	for (i = 0; i < CNTR_CNT; i++) {
+		cntr = rdm->util_ep.cntrs[i];
+		if (!cntr)
+			continue;
+
+		if (cntr->domain == &domain->util_domain) {
+			ofi_atomic_dec32(&cntr->domain->ref);
+			ofi_atomic_inc32(&subdomain->util_domain.ref);
+			cntr->domain = &subdomain->util_domain;
+		}
+	}
+}
+
+int xnet_rdm_resolve_domains(struct xnet_rdm *rdm)
+{
+	int ret;
+	struct fid_domain *subdomain_fid;
+	struct xnet_domain *subdomain;
+	struct xnet_domain *domain = container_of(rdm->util_ep.domain,
+							struct xnet_domain,
+							util_domain);
+	subdomain = find_subdomain(rdm);
+	if (!subdomain) {
+		ret = fi_domain(&domain->util_domain.fabric->fabric_fid,
+				domain->subdomain_info,
+				&subdomain_fid, NULL);
+		if (ret)
+			return ret;
+
+		subdomain = container_of(subdomain_fid, struct xnet_domain,
+					 util_domain.domain_fid);
+		ret = fid_list_insert2(&domain->subdomain_list,
+				       &domain->subdomain_list_lock,
+				       &subdomain_fid->fid);
+		if (ret)
+			fi_close(&subdomain_fid->fid);
+
+		set_subdomain(rdm, domain, subdomain);
+
+		ofi_rbmap_foreach(domain->util_domain.mr_map.rbtree,
+				  domain->util_domain.mr_map.rbtree->root,
+				  xnet_reg_subdomain_mr, subdomain);
+	}
```

Comment:
If I understand correctly, `find_subdomain()` handles the scenario that the cq(s)/cntr(s) bound to this rdm ep may have previously been bound to another rdm ep that has already been enabled (and thus already got a subdomain assigned), in which case the two rdm eps will share the same subdomain since they belong to the same completion group. However, there is the possibility that some of cq(s)/cntr(s) bound to this rdm ep may still be associated with the multiplex domain. So `set_subdomain()` needs also be called from the `else` branch (which is not present here).